### PR TITLE
feat: auto-import for sway

### PR DESCRIPTION
**Pull request by sdankel**
## Description

Closes https://github.com/FuelLabs/sway/issues/4782

Notes

This implementation relies on the diagnostics that are provided in the context of a code action request. Conveniently, only the diagnostics related to the symbol in the request are provided in the context. Since we also need the name of the symbol to look up, I added a `DiagnosticData` type to the metadata that is published with diagnostics. For the errors that indicate a missing symbol, this data is attached, and then used to look for possible imports when the code action request handler runs.

- Added `ProgramTypeKeyword` symbol kind to distinguish them from other keywords
- Added `submodules_recursive` to lexed and parsed modules because LSP wasn't collecting tokens inside of nested submodules. This implementation comes from typed module, I just made it generic via the `HasSubmodules` and `HasModule` traits.
- Added `TyIncludeStatement` to the compiler, and span and mod_name to `IncludeStatement` to populate it. This was needed to identify where the end of the last `mod` statement is in the file.
- Added `span` to `UseStatment` and `TyUseStatement` so that we can replace an entire Group statement with the new one. e.g. `use a::b::{C, D, E};`
- Added `CallPath` to several of the typed tokens that were missing it.

### Testing

- integration tests for each type of item that can be imported
- unit tests for the location of the generated import in the file

### Limitations

I found several bugs / edge cases where auto-import doesn't work:
- enum variants: https://github.com/FuelLabs/sway/issues/5188 
- std-lib functions: https://github.com/FuelLabs/sway/issues/5191
- std-lib constants: https://github.com/FuelLabs/sway/issues/5192

## Checklist

- [ ] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: sway-core/src/semantic_analysis/ast_node/declaration/trait.rs [hunk 1]
```diff
@@ -210,14 +210,15 @@ impl TyTraitDecl {
...
-            name,
+            name: name.clone(),
             type_parameters: new_type_parameters,
             self_type: self_type_param,
             interface_surface: new_interface_surface,
             items: new_items,
             supertraits,
             visibility,
             attributes,
+            call_path: CallPath::from(name.clone()).to_fullpath(ctx.namespace),
```
Comment (cr-fuel) [inline sway-core/src/semantic_analysis/ast_node/declaration/trait.rs:221]: I think this `name.clone()` should be just `name`

File: sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs [hunk 2]
```diff
@@ -279,6 +283,7 @@ fn use_tree_to_use_statements(
             };
             ret.push(UseStatement {
                 call_path: path.clone(),
+                span: item_span.clone(),
```
Comment (cr-fuel) [inline sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs:286]: The `.clone()` is not needed here

File: sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs [hunk 3]
```diff
@@ -292,6 +297,7 @@ fn use_tree_to_use_statements(
...
```
Comment (cr-fuel) [inline sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs:300]: The `.clone()` is not needed here

File: sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs [hunk 4]
```diff
@@ -300,14 +306,15 @@ fn use_tree_to_use_statements(
         UseTree::Glob { .. } => {
             ret.push(UseStatement {
                 call_path: path.clone(),
+                span: item_span.clone(),
```
Comment (cr-fuel) [inline sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs:309]: The `.clone()` is not needed here

File: sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs [hunk 5]
```diff
@@ -300,14 +306,15 @@ fn use_tree_to_use_statements(
...
                 import_type: ImportType::Star,
                 is_absolute,
                 alias: None,
             });
         }
         UseTree::Path { prefix, suffix, .. } => {
             path.push(prefix);
-            use_tree_to_use_statements(*suffix, is_absolute, path, ret);
+            use_tree_to_use_statements(*suffix, is_absolute, path, ret, item_span.clone());
```
Comment (cr-fuel) [inline sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs:317]: The `.clone()` is not needed here

**Review by cr-fuel — CHANGES_REQUESTED** [review]
Just a few minor notes, otherwise LGTM

File: sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs (hunk 2 above)
Comment (sdankel) [reply sway-core/src/transform/to_parsed_lang/convert_parse_tree.rs]: Good catch, I'm surprised clippy didn't find these!

**Review by cr-fuel — DISMISSED** [review]

File: sway-core/src/language/ty/declaration/function.rs [hunk 6]
```diff
@@ -240,6 +242,7 @@ impl TyFunctionDecl {
             },
             implementing_type: None,
             span,
+            call_path: CallPath::from(Ident::new_no_span("foo".into())),
```
Comment (JoshuaBatty) [inline sway-core/src/language/ty/declaration/function.rs:245]: Perhaps a comment here might be helpful to help readers understand why an Ident with no span named foo is used?

File: sway-core/src/semantic_analysis/ast_node/declaration/function.rs [hunk 7]
```diff
@@ -354,6 +358,7 @@ fn test_function_selector_behavior() {
             },
         ],
         span: Span::dummy(),
+        call_path: CallPath::from(Ident::new_no_span("foo".into())),
```
Comment (JoshuaBatty) [inline sway-core/src/semantic_analysis/ast_node/declaration/function.rs:361]: again here.

File: sway-lsp/src/capabilities/code_actions/diagnostic/auto_import.rs [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+                        }
+                        TyDecl::TraitDecl(decl) => {
+                            let trait_decl = ctx.engines.de().get_trait(&decl.decl_id);
+                            let call_path = trait_decl.call_path.to_import_path(&namespace);
+                            Some(call_path)
+                        }
+                        _ => None,
+                    };
+                }
+                Some(TypedAstToken::TypedFunctionDeclaration(ty_decl)) => {
```
Comment (JoshuaBatty) [inline sway-lsp/src/capabilities/code_actions/diagnostic/auto_import.rs:117]: maybe this is cleaner? 
```suggestion
                Some(TypedAstToken::TypedFunctionDeclaration(TyFunctionDecl{call_path, ..}))
                | Some(TypedAstToken::TypedConstantDeclaration(TyConstantDecl{call_path, ..}))
                | Some(TypedAstToken::TypedTypeAliasDeclaration(TyTypeAliasDecl{call_path, ..})) => {
                    let call_path = ty_decl.call_path.to_import_path(&namespace);
                    Some(call_path)
                }
```

**Review by JoshuaBatty — COMMENTED** [review]
Great work @sdankel . Lot's of nice cleanups along the way. Just a couple of nits.

File: sway-core/src/language/ty/declaration/function.rs (hunk 6 above)
Comment (sdankel) [reply sway-core/src/language/ty/declaration/function.rs:245]: Moved it to an `Ident::dummy()` method since this is used elsewhere.

**Comment by JoshuaBatty** [comment]
Just tested this locally with the test example. Seems each import into the `deep_mod` use statement added a new line leading to a large gap between it and the next `use` statement. Are we able to avoid this?

Screenshot 2023-10-17 at 10 59 11 am

**Comment by sdankel** [comment]
> Just tested this locally with the test example. Seems each import into the `deep_mod` use statement added a new line leading to a large gap between it and the next `use` statement. Are we able to avoid this?
> 
> Screenshot 2023-10-17 at 10 59 11 am

Good catch! Fixed.

**Review by JoshuaBatty — APPROVED** [review]
🚀

File: sway-core/src/language/module.rs [hunk 9]
```diff
@@ -5,3 +5,72 @@ use sway_types::Ident;
...
+    fn module(&self) -> &T;
+}
+
+pub trait HasSubmodules<E>
+where
+    E: HasModule<Self>,
+    Self: Sized,
+{
+    /// Returns the submodules of this module.
+    fn submodules(&self) -> &Vec<(ModName, E)>;
```
Comment (cr-fuel) [inline sway-core/src/language/module.rs:24]: Any reason to return a `&Vec<(ModName, E)>` instead of `&[(ModName, E)]` ?

Comment (sdankel) [reply sway-core/src/language/module.rs]: No reason, updated

**Review by JoshuaBatty — APPROVED** [review]

File: sway-core/src/language/ty/declaration/trait.rs [hunk 10]
```diff
@@ -30,6 +30,7 @@ pub struct TyTraitDecl {
     pub supertraits: Vec<parsed::Supertrait>,
     pub visibility: Visibility,
     pub attributes: transform::AttributesMap,
+    pub call_path: CallPath,
```
Comment (IGI-111) [inline sway-core/src/language/ty/declaration/trait.rs:33]: This needs a better, more descriptive name, as it seems you want to guarantee that it'll be a full path to the trait decl and not just any `CallPath`.

**Review by IGI-111 — COMMENTED** [review]
LGTM but I do insist on that name change as people seem to be easily confused that `CallPath` values from the typed AST are not guaranteed to be valid top level fullpaths, which this needs to be.

**Comment by sdankel** [comment]
> LGTM but I do insist on that name change as people seem to be easily confused about the fact that `CallPath` values from the typed AST are not guaranteed to be valid top level fullpaths, which this explicitly needs to be as I understand the usage.

It doesn't need to be a full path actually, since I end up calling `to_fullpath` on it anyway. It's just there for symmetry with the other types (TyEnumDecl, TyStructDecl) to make the code more consistent.

**Comment by IGI-111** [comment]
Ah never mind that then

**Review by IGI-111 — APPROVED** [review]
